torch/_inductor/fx_passes/b2b_gemm.py

```diff
@@ -0,0 +1,188 @@
+        1,
+    )
+
+b2b_gemm_template = TritonTemplate(
+    name="b2b_gemm",
+    grid=b2b_gemm_grid,
+    debug=True,
+    source=r"""
+{{def_kernel("A", "B", "C")}}
+    M = {{size("A", 0)}}
+    # N = {{size("A", 1)}}
+    O = {{size("C", 0)}}
+    # P = {{size("C", 1)}}
+
+    stride_am = {{stride("A", 0)}}
+    stride_an = {{stride("A", 1)}}
+    stride_bn = {{stride("B", 0)}}
+    stride_bo = {{stride("B", 1)}}
+    stride_co = {{stride("C", 0)}}
+    stride_cp = {{stride("C", 1)}}
+
+    # A's row block for this thread
+    row_block_id = tl.program_id(axis=0)
+
+    # divide B's columns (and C's rows)
+    num_col_block = tl.cdiv(O, COL_BLOCK_SIZE)
+
+    # offsets (TODO: handle the non-divisible case)
+    offs_row = row_block_id * ROW_BLOCK_SIZE + tl.arange(0, ROW_BLOCK_SIZE)
+    offs_col = tl.arange(0, COL_BLOCK_SIZE)  # to be updated in the loop
+
+    # accumulator
+    acc = tl.zeros((ROW_BLOCK_SIZE, P), dtype=tl.float16)
+
+    a_ptrs = A + (offs_row[:, None] * stride_am + tl.arange(0, N)[None, :] * stride_an)
+    a = tl.load(a_ptrs)
+
+    for _ in range(num_col_block):
+
+        b_ptrs = B + (tl.arange(0, N)[:, None] * stride_bn + offs_col[None, :] * stride_bo)
+        b = tl.load(b_ptrs)
+
+        c_ptrs = C + (offs_col[:, None] * stride_co + tl.arange(0, P)[None, :] * stride_cp)
+        c = tl.load(c_ptrs)
+
+        # computation (TODO: floating point errors)
+        acc += tl.dot(tl.dot(a, b, out_dtype=tl.float16), c, out_dtype=tl.float16)
+
+        # update offsets
+        offs_col += COL_BLOCK_SIZE
+
+    # store
+    idx_m = offs_row[:, None]
+    idx_p = tl.arange(0, P)[None, :]
+    mask = (idx_m < M) & (idx_p < P)
+
+    {{store_output(("idx_m", "idx_p"), "acc", "mask")}}
+""",
+)
+
+def can_apply_b2b_gemm(mat1, mat2, mat3) -> bool:
+    if not (("val" in mat1.meta) and ("val" in mat2.meta) and ("val" in mat3.meta)):
+        return False
+    mat1 = mat1.meta["val"]
+    mat2 = mat2.meta["val"]
+    mat3 = mat3.meta["val"]
+    if not (mat1.is_cuda and mat2.is_cuda and mat3.is_cuda):
+        return False
+    if not ((len(mat1.shape) == 2) and (len(mat2.shape) == 2) and (len(mat3.shape) == 2)):
+        return False
+    if not ((mat1.shape[1] == mat2.shape[0]) and (mat2.shape[1] == mat3.shape[0])):
+        return False
+    # TODO: change to a real-check for size restrictions (may consider hardware limit?)
+    m, n, o, p = mat1.shape[0], mat1.shape[1], mat3.shape[0], mat3.shape[1]
+    m_ok = (m % 128 == 0) and (m > 128)
+    n_ok = (n == 32)
+    o_ok = (o % 128 == 0) and (o > 128)
+    p_ok = (p == 32)
+    return (m_ok and n_ok and o_ok and p_ok)
+
+def tuned_b2b_gemm(mat1, mat2, mat3, *, layout=None):
+    layout = FixedLayout(
+        mat1.get_device(), mat1.get_dtype(), [mat1.shape[0], mat3.shape[1]]
+    )
+    choices = []
+    # TODO: change N and P to non-constexpr
+    # Note: the only reason why N and P are hardcoded is because in Triton tl.arange(0, N) only works for tl.constexpr
```

Comment:
Do you have a plan to fix that? (I actually have no idea what's involved)

Reply:
N is used as index tensors' size in the Triton kernel. So there are two ways to fix it:
(1) if there is a way to get non-constexpr-size index tensors, we can use that;
(2) if there is no way to get non-constexpr-size index tensors, we can further divide the matrix into smaller blocks.

Reply:
Fixed.